File: cmd/tonelab-cli/steps.go
package main

import (
	"encoding/json"
	"fmt"
	"strings"

	"tonelab/backend/app"
)

// describeStep turns one tool call into a sentence, as the window does,
// with a short account of what came back. The tool names and JSON are
// ours, not the user's, and a transcript that reads like a log is one
// nobody reads.
func describeStep(step app.JournalStep) (call, result string) {
	args := map[string]any{}
	json.Unmarshal([]byte(step.Arguments), &args)
	track := fmt.Sprint(args["track_id"])
	name := fmt.Sprint(args["param_name"])
	switch step.Tool {
	case "list_tracks":
		call = "Looked up the tracks"
	case "get_param":
		call = fmt.Sprintf("Read %s on track %s", name, track)
	case "set_param":
		call = fmt.Sprintf("Set %s on track %s to %s", name, track, format(args["value"]))
	case "undo":
		call = "Asked the DAW to undo"
	case "list_fx":
		call = fmt.Sprintf("Looked up the effects on track %s", track)
	case "find_params":
		call = fmt.Sprintf("Searched track %s for %q", track, fmt.Sprint(args["query"]))
	case "get_fx_param":
		call = fmt.Sprintf("Read %s on track %s", fxTarget(step.Outcome, args), track)
	case "set_fx_param":
		call = fmt.Sprintf("Set %s on track %s to %s", fxTarget(step.Outcome, args), track, format(args["value"]))
	case "search":
		call = fmt.Sprintf("Searched the web for %q", fmt.Sprint(args["query"]))
	case "fetch_page":
		call = fmt.Sprintf("Read %s", fmt.Sprint(args["url"]))
	default:
		call = step.Tool
	}
	return call, summarize(step)
}

// fxTarget names the effect and parameter from the outcome, where the
// backend put the names a user recognises; the arguments carry only ids.
func fxTarget(outcome string, args map[string]any) string {
	var reply struct {
		Value struct {
			FXName string `json:"fx_name"`
			Name   string `json:"name"`
		} `json:"value"`
	}
	if json.Unmarshal([]byte(outcome), &reply) == nil && reply.Value.Name != "" {
		return reply.Value.FXName + " / " + reply.Value.Name
	}
	return fmt.Sprintf("effect %v parameter %v", args["fx_id"], args["param_id"])
}

// summarize is the one line under a call: a count, a value, or the
// refusal. The full outcome stays in the history for anyone who wants it.
func summarize(step app.JournalStep) string {
	var outcome map[string]any
	if json.Unmarshal([]byte(step.Outcome), &outcome) != nil {
		return ""
	}
	if failure, ok := outcome["error"].(map[string]any); ok {
		return fmt.Sprint(failure["message"])
	}
	value := outcome["value"]
	switch v := value.(type) {
	case float64, bool:
		return format(v)
	case []any:
		return plural(len(v), "result")
	case map[string]any:
		if matches, ok := v["matches"].([]any); ok {
			return plural(len(matches), "match")
		}
		if fx, ok := v["fx"].([]any); ok {
			return plural(len(fx), "effect")
		}
		if confirmed, ok := v["confirmed"]; ok {
			return "confirmed " + format(confirmed)
		}
		if would, ok := v["would"].(string); ok {
			return "would " + would
		}
		if note, ok := v["note"].(string); ok {
			return note
		}
		if reading, ok := v["value"]; ok {
			return format(reading)
		}
		if text, ok := v["text"].(string); ok {
			return fmt.Sprintf("%d characters", len(text))
		}
	}
	return strings.TrimSpace(fmt.Sprint(value))
}

func plural(n int, noun string) string {
	if n == 1 {
		return "1 " + noun
	}
	if strings.HasSuffix(noun, "ch") {
		return fmt.Sprintf("%d %ses", n, noun)
	}
	return fmt.Sprintf("%d %ss", n, noun)
}

File: frontend/src/lib/steps.ts
// Translated rather than printed. The tool names and JSON are ours, not
// the user's, and a history that reads like a log is one nobody reads.
import { format, parseJSON } from "./format";

export type StepKind = "list" | "read" | "set" | "undo" | "other";

// The kind of step, so a history can be scanned rather than read.
export function kindOf(tool: string): StepKind {
    switch (tool) {
        case "list_tracks": case "list_fx": case "find_params": case "search": return "list";
        case "get_param": case "get_fx_param": case "fetch_page": return "read";
        case "set_param": case "set_fx_param": return "set";
        case "undo": return "undo";
        default: return "other";
    }
}

export function describeStep(tool: string, args: string, outcome: string, failed: boolean): string {
    const parsed = parseJSON(args);
    const track = parsed.track_id ?? "";
    const name = parsed.param_name ?? "";

    let said: string;
    switch (tool) {
        case "list_tracks": said = "Looked up the tracks"; break;
        case "get_param": said = `Read ${name} on track ${track}`; break;
        case "set_param": said = `Set ${name} on track ${track} to ${format(parsed.value)}`; break;
        case "undo": said = "Asked the DAW to undo"; break;
        case "list_fx": said = `Looked up the effects on track ${track}`; break;
        case "find_params": said = `Searched track ${track} for "${parsed.query ?? ""}"`; break;
        case "search": said = `Searched the web for "${parsed.query ?? ""}"`; break;
        case "fetch_page": said = `Read ${parsed.url ?? "a page"}`; break;
        case "get_fx_param": said = `Read ${fxTarget(outcome, parsed)} on track ${track}`; break;
        case "set_fx_param": said = `Set ${fxTarget(outcome, parsed)} on track ${track} to ${format(parsed.value)}`; break;
        default: said = tool;
    }
    if (failed) {
        const reason = parseJSON(outcome).error?.message;
        return `${said}, refused${reason ? `: ${reason}` : ""}`;
    }
    return said;
}

// The call carries positions and the outcome carries names; a person wants
// the names, so they are taken from the outcome when the call succeeded.
function fxTarget(outcome: string, parsed: Record<string, unknown>): string {
    const value = parseJSON(outcome).value ?? {};
    if (value.fx_name && value.name) return `${value.name} on ${value.fx_name}`;
    return `effect ${parsed.fx_id ?? "?"} parameter ${parsed.param_id ?? "?"}`;
}
